### PR TITLE
Add label topwatchers if number of comments or positive reactions >= 5

### PR DESCRIPTION
**Pull request by sowbiba**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | If a comment is added to an issue, and number of comments other than author or PS employees reached the defined threshold, add a label "Topwatchers" to the issue
| Type?             | new feature
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes {paste the issue here}.
| How to test?      | -

### CODE REVIEW

File: src/ruleFinder/IssueCommentsRuleFinder.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+     * @param {Context} context
+     *
+     * @returns {Promise<*>}
+     *
+     * @public
+     */
+  async findRules(context) {
+    const rules = [];
+
+    if (Utils.contextHasAction(context, 'created')) {
```
Comment (PierreRambaud) [inline src/ruleFinder/IssueCommentsRuleFinder.js:54]: Seems to be a huge if condition

File: src/issue-comments-data-provider.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      owner,
+      repo,
+    });
+
+    let positiveReactionsCount = 0;
+    if (excludeIssueOwner) {
+      excludedUsers.push(owner)
+    }
+
+    for (let index = 0; index < data.length; index += 1) {
```
Comment (PierreRambaud) [inline src/issue-comments-data-provider.js:97]: Why not using ` data.forEach`  instead?

**Review by Progi1984 — COMMENTED** [review]
A small comment

File: src/rule.js [hunk 3]
```diff
@@ -228,6 +228,13 @@ const Rules = {
    * AND add the label “TBS” to the Issue
    * AND remove assignee if there is one
    */
+  M1: 'M1',
+  /*
```
Comment (PierreRambaud) [inline src/rule.js:232]: Comments shouldn't be above the task? :thinking:

File: src/ruleFinder/IssueCommentsRuleFinder.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  /**
+   * Parse a github URL to extract Issue / Pull Request informations
+   *
+   * @param {string} url
+   *
+   * @returns {object}
+   */
+  parseUrlForData(url) {
```
Comment (PierreRambaud) [inline src/ruleFinder/IssueCommentsRuleFinder.js:102]: Duplicate from rule-computer

File: src/ruleFinder/IssueCommentsRuleFinder.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+  /**
+     * Try to find whether webhook context matches an Issue rule requirements.
+     *
+     * @param {Context} context
+     *
+     * @returns {Promise<*>}
+     *
+     * @public
+     */
```
Comment (PierreRambaud) [inline src/ruleFinder/IssueCommentsRuleFinder.js:50]:
```suggestion
  /**
   * Try to find whether webhook context matches an Issue rule requirements.
   *
   * @param {Context} context
   *
   * @returns {Promise<*>}
   *
   * @public
   */
```

File: src/ruleFinder/IssueCommentsRuleFinder.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+const Rule = require('../rule');
+const Utils = require('./Utils');
+
+module.exports = class IssueCommentsRuleFinder {
+  /**
+     * @param config
+     * @param {IssueCommentsDataProvider} issueCommentsDataProvider
+     * @param {ConfigProvider} configProvider
+     * @param {Logger} logger
+     */
```
Comment (PierreRambaud) [inline src/ruleFinder/IssueCommentsRuleFinder.js:34]:
```suggestion
  /**
   * @param config
   * @param {IssueCommentsDataProvider} issueCommentsDataProvider
   * @param {ConfigProvider} configProvider
   * @param {Logger} logger
   */
```

File: src/ruleFinder/IssueCommentsRuleFinder.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    const rules = [];
+
+    if (!Utils.contextHasAction(context, 'created')) {
+      return [];
+    }
+
+    const {issue} = context.payload;
+    const repositoryConfig = this.configProvider.getRepositoryConfigFromIssue(this.config, issue);
+    // No need to go further if comments is lowest than threshold
+    if (issue.comments >= repositoryConfig.topwatchersThreshold) {
```
Comment (PierreRambaud) [inline src/ruleFinder/IssueCommentsRuleFinder.js:61]:
```suggestion
    if (issue.comments < repositoryConfig.topwatchersThreshold) {
      return [];
    }
```

File: src/rules/M1.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+const Rule = require('./Rule.js');
+const Utils = require('../ruleFinder/Utils');
+
+module.exports = class D1 extends Rule {
```
Comment (PierreRambaud) [inline src/rules/M1.js:28]:
```suggestion
module.exports = class M1 extends Rule {
```

**Review by PierreRambaud — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by Progi1984** [comment]
Thanks @sowbiba

**Pull request merged**